connman/src/storage.c

```diff
@@ -794,11 +796,53 @@ static void keyfile_cleanup(void)
 	keyfile_hash = NULL;
 }
 
+static int validate_path(const char *pathname)
+{
+	char resolved_path[PATH_MAX] = { 0 };
+
+	/*
+	 * realpath() expands all symbolic links and references and handles
+	 * NULL pathname.
+	 */
+	if (!realpath(pathname, resolved_path)) {
+		switch (errno) {
+		/*
+		 * File does not exist, which is the case when saving. Verify
+		 * the path to avoid using a broken symlink or unaccessible
+		 * file.
+		 */
+		case ENOENT:
+			DBG("path %s is not readable/has missing symlink (%s)",
+						pathname, resolved_path);
+			break;
```

Comment:
I wonder should we care that 'resolved_path' content is undefined here / after break unless we are dealing with realpath() with gnu extensions? Is it even possible to know which way it is?

Reply:
That is a reasonably good question; musl's realpath() doesn't seem to implement the GNU extension.

Reply:
What might be sufficiently close to gnu ENOENT behavior we are after could be something like:
if realpath(parent(pathname)) succeeds, append basename(pathname) back in and do access()/stat()/whatnot checks on the result (in case it happens to be something weird like a broken link).

Though given that afaik we are mostly interested in not having symlinks on the way, that openat() loop starts to look like something simple and with well defined behavior to me...

Reply:
Files are saved surprisingly often so with every file save that allocation caused by splitting of the path seems to be a bit too much. Anyways, `g_file_set_contents()` does not allow saving a file to a path with a broken symlink...

Reply:
Well, I'd assume internally realpath() is basically something like that and then some more, so in all likelihood custom solution that bails out the moment there is a symlink imight actually be the lighter option.

But for our purposes relying on gnu extended realpath() works, so I'll leave it up to you whether stricter standards compliance is needed or not.

Reply:
Yeah, this is by far the simplest implementation, but I suppose the performance difference is very minor (esp. if you implement the splitting with strtok/similar instead of g_strsplit).

If one wants to go with the non-racy version, that would require re-implementing quite a few helpers (even mkstemp() doesn't have a *at() version).

I'm cool with current implementation, even with the slight race it leaves behind.